### ISSUE TITLE
cellbender -> pytorch -> disk full, but it is not!

### ISSUE THREAD

**Issue by smoe** [nf-core/scdownstream#58]
### Description of the bug

Ok, I get a "disk full" error. With large single-cell datasets this does not sound completely unexpected, but it is also why I am in our HPC environment, which from what "df -h" keeps telling me has many hundreds of terabytes at my disposal. The only exception is $(HOME) with just some 35GB.  /tmp seems to be fine, too (not memory mapped). A disk /scratch has a local 1TB ssd waiting that nobody knows about. The work directory is on that huge /data one.

For better compatibility with singularity, @nictru kindly suggested adding
```
singularity.runOptions = '--no-mount tmp --writable-tmpfs'
```
to nextflow.config. Is that in conflict with that pytorch magic?

I have an environment setting for Singularity
```
$ env|grep -i singu
SINGULARITY_BIND=/data
```
which I can possibly extend to my home directory and then have the singularity.runOptions changed back to the defaults?

### Command used and terminal output

```console
(base) [sm718@haumea1 4392e73f083be38edd2f0f80698751]$ cat .command.err 
cellbender:remove-background: Command:
cellbender remove-background --input JM_24h_unified.h5ad --output JM_24h_cellbender.h5
cellbender:remove-background: CellBender 0.3.0
cellbender:remove-background: (Workflow hash b4603011ef)
cellbender:remove-background: 2024-07-04 11:11:36
cellbender:remove-background: Running remove-background
cellbender:remove-background: Loading data from JM_24h_unified.h5ad
cellbender:remove-background: AnnData format
cellbender:remove-background: Found `.layers['counts']`. Using for count data.
cellbender:remove-background: Only 6834 barcodes were found.
This suggests the matrix was prefiltered.
CellBender requires a raw, unfiltered [Barcodes, Genes] matrix.
cellbender:remove-background: WARNING: Only 6834 barcodes in the input file. Ensure this is a raw (unfiltered) file with all barcodes, including the empty droplets.
cellbender:remove-background: Features in dataset: 89916 NA
cellbender:remove-background: Trimming features for inference.
cellbender:remove-background: 55800 features have nonzero counts.
cellbender:remove-background: Prior on counts for cells is 22156
cellbender:remove-background: Prior on counts for empty droplets is 10829
cellbender:remove-background: Excluding 3138 features that are estimated to have <= 0.1 background counts in cells.
cellbender:remove-background: Including 52662 features in the analysis.
cellbender:remove-background: Trimming barcodes for inference.
cellbender:remove-background: Excluding barcodes with counts below 5414
cellbender:remove-background: Using 464 probable cell barcodes, plus an additional 2673 barcodes, and 2981 empty droplets.
cellbender:remove-background: Largest surely-empty droplet has 11967 UMI counts.
cellbender:remove-background: Attempting to unpack tarball "ckpt.tar.gz" to /tmp/tmp_ymlxgn8
cellbender:remove-background: No saved checkpoint.
cellbender:remove-background: No checkpoint loaded.
cellbender:remove-background: Running inference...
cellbender:remove-background: [epoch 001]  average training loss: 23449.0117
cellbender:remove-background: [epoch 002]  average training loss: 22924.1592  (34.2 seconds per epoch)
cellbender:remove-background: Will checkpoint every 13 epochs
cellbender:remove-background: [epoch 003]  average training loss: 22145.3729
cellbender:remove-background: [epoch 004]  average training loss: 21626.0684
cellbender:remove-background: [epoch 005]  average training loss: 21037.9555
cellbender:remove-background: [epoch 005] average test loss: 22359.1018
cellbender:remove-background: [epoch 006]  average training loss: 20722.9589
cellbender:remove-background: [epoch 007]  average training loss: 20325.7407
cellbender:remove-background: [epoch 008]  average training loss: 20033.0595
cellbender:remove-background: [epoch 009]  average training loss: 19803.3412
cellbender:remove-background: [epoch 010]  average training loss: 19571.5222
cellbender:remove-background: [epoch 010] average test loss: 20960.7742
cellbender:remove-background: [epoch 011]  average training loss: 19405.0574
cellbender:remove-background: [epoch 012]  average training loss: 19188.8284
cellbender:remove-background: [epoch 013]  average training loss: 19022.1505
cellbender:remove-background: Saving a checkpoint...
cellbender:remove-background: Could not save checkpoint
cellbender:remove-background: Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/serialization.py", line 379, in save
    _save(obj, opened_zipfile, pickle_module, pickle_protocol)
  File "/opt/conda/lib/python3.7/site-packages/torch/serialization.py", line 604, in _save
    zip_file.write_record(name, storage.data_ptr(), num_bytes)
OSError: [Errno 28] No space left on device

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/cellbender/remove_background/checkpoint.py", line 115, in save_checkpoint
    torch.save(model_obj, filebase + '_model.torch')
  File "/opt/conda/lib/python3.7/site-packages/torch/serialization.py", line 380, in save
    return
  File "/opt/conda/lib/python3.7/site-packages/torch/serialization.py", line 259, in __exit__
    self.file_like.write_end_of_file()
RuntimeError: [enforce fail at inline_container.cc:319] . unexpected pos 5469568 vs 5469456

terminate called after throwing an instance of 'c10::Error'
  what():  [enforce fail at inline_container.cc:319] . unexpected pos 5469568 vs 5469456
frame #0: c10::ThrowEnforceNotMet(char const*, int, char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, void const*) + 0x64 (0x2b02ea7e7cf4 in /opt/conda/lib/python3.7/site-packages/torch/lib/libc10.so)
frame #1: <unknown function> + 0x2bb7329 (0x2b02e41c0329 in /opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so)
frame #2: mz_zip_writer_add_mem_ex_v2 + 0x603 (0x2b02e41ba573 in /opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so)
frame #3: caffe2::serialize::PyTorchStreamWriter::writeRecord(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, void const*, unsigned long, bool) + 0x113 (0x2b02e41c2e23 in /opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so)
frame #4: caffe2::serialize::PyTorchStreamWriter::writeEndOfFile() + 0x49b (0x2b02e41c378b in /opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so)
frame #5: caffe2::serialize::PyTorchStreamWriter::~PyTorchStreamWriter() + 0x17e (0x2b02e41c3a0e in /opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so)
frame #6: <unknown function> + 0x6b81fb (0x2b02e0eec1fb in /opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_python.so)
frame #7: <unknown function> + 0x346b02 (0x2b02e0b7ab02 in /opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_python.so)
frame #8: <unknown function> + 0x3478ff (0x2b02e0b7b8ff in /opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_python.so)
frame #9: <unknown function> + 0xe5698 (0x56063f15c698 in /opt/conda/bin/python3.7)
frame #10: <unknown function> + 0xe6038 (0x56063f15d038 in /opt/conda/bin/python3.7)
frame #11: <unknown function> + 0xe604e (0x56063f15d04e in /opt/conda/bin/python3.7)
frame #12: _PyEval_EvalFrameDefault + 0x5bcf (0x56063f2341af in /opt/conda/bin/python3.7)
frame #13: _PyEval_EvalCodeWithName + 0x33d (0x56063f174ccd in /opt/conda/bin/python3.7)
frame #14: _PyFunction_FastCallKeywords + 0x320 (0x56063f1a6570 in /opt/conda/bin/python3.7)
frame #15: <unknown function> + 0x191de8 (0x56063f208de8 in /opt/conda/bin/python3.7)
frame #16: _PyEval_EvalFrameDefault + 0x16bb (0x56063f22fc9b in /opt/conda/bin/python3.7)
```

### System information

      N E X T F L O W
      version 24.04.2 build 5914
      created 29-05-2024 06:19 UTC (08:19 MESZ)

HPC
slurm
Singularity
CentOS Linux
-r dev

**Comment by nictru** [comment]
Could you attach the `.nextflow.log` of this run?

**Comment by nictru** [comment]
Ah I was able to replicate the issue, no need for the log. Will add the fix to #55 once I have it

**Comment by nictru** [comment]
I have not been able to find out why exactly the `No space left on device` occurs, but was able to relocate the `TMPDIR` to the process workdir, so that we have more control over what is happening.
